pkg/v1/sdk/capabilities/discovery/cluster_gvr.go

```diff
@@ -35,94 +58,236 @@ func (q *QueryGVR) Name() string {
 }
 
 // WithVersions checks if an API group with all the specified versions exist.
+// This method can be omitted to query any version.
 func (q *QueryGVR) WithVersions(versions ...string) *QueryGVR {
 	q.versions = versions
 	return q
 }
 
 // WithResource checks if an API group with the specified resource exists.
-// WithVersions needs to be used before calling this function.
+// This method can be omitted to query any resource.
 func (q *QueryGVR) WithResource(resource string) *QueryGVR {
-	q.resource = resource
+	q.resource.set(resource)
 	return q
 }
 
-// Run discovery
+// Run discovery.
 func (q *QueryGVR) Run(config *clusterQueryClientConfig) (bool, error) {
-	if q.group != "" && q.resource != "" && len(q.versions) == 0 {
-		return false, fmt.Errorf("cannot check for group and resource existence without version info; use WithVersion method")
+	if err := q.validate(config); err != nil {
+		return false, fmt.Errorf("validating query: %w", err)
+	}
+
+	var unmatched []string
+
+	switch {
+	case !q.hasVersion():
+		unmatchedGroupResource, err := q.unmatchedGroupResource(config)
+		if err != nil {
+			return false, fmt.Errorf("discovering unmatched GroupResource: %w", err)
+		}
+		if unmatchedGroupResource != "" {
+			unmatched = append(unmatched, unmatchedGroupResource)
+		}
+	case q.hasVersion() && q.resource.String == "":
+		unmatchedGroupVersions, err := q.unmatchedGroupVersions(config)
+		if err != nil {
+			return false, fmt.Errorf("discovering unmatched GroupVersions: %w", err)
+		}
+		unmatched = append(unmatched, unmatchedGroupVersions...)
+	case q.hasVersion() && q.resource.String != "":
+		unmatchedGVRs, err := q.unmatchedGroupVersionResources(config)
+		if err != nil {
+			return false, fmt.Errorf("discovering unmatched GroupVersionResources: %w", err)
+		}
+		unmatched = append(unmatched, unmatchedGVRs...)
 	}
 
-	var gvrs []schema.GroupVersionResource
-	if len(q.versions) > 0 {
-		for _, v := range q.versions {
-			gvrs = append(gvrs, schema.GroupVersionResource{Group: q.group, Version: v, Resource: q.resource})
+	q.unmatchedGVRs = unmatched
+	return len(unmatched) == 0, nil
+}
+
+func (q *QueryGVR) validate(cfg *clusterQueryClientConfig) error {
+	if cfg == nil {
+		return fmt.Errorf("clusterQueryClientConfig must not be nil")
+	}
+
+	if q.resource.IsSet && q.resource.String == "" {
+		return ErrInvalidWithResourceMethodArgument
+	}
+
+	if q.versions != nil && !q.hasVersion() {
+		return ErrInvalidWithVersionsMethodArgument
+	}
+
+	return nil
+}
+
+func (q *QueryGVR) hasVersion() bool {
+	var has bool
+	for _, ver := range q.versions {
+		ver = strings.TrimSpace(ver)
+		if ver != "" {
+			has = true
 		}
-	} else {
-		gvrs = append(gvrs, schema.GroupVersionResource{Group: q.group})
+	}
+	return has
+}
```

Comment:
I don't think this function is complete. Specifying `WithVersion("v1", "")` is an error in my opinion. But this will return true because it only sets true, never false.

If we've decided that the empty string is an invalid value for version, then that should always be true regardless of the position of the empty string.

Reply:
After some thought, I agree with you, @jmoroski.

What the code currently does is if someone really wanted to sneak in an empty version, as long as it had one other version, I'd let the query go through. In the end, the query would not error, but would return `Found: false` because it could not find the empty version. 

However, allowing this loophole is inconsistent with our design decision and makes understanding our API more complicated than it needs to be.

I'll update the logic to not allow any empty string versions, even if there are valid versions listed.